src/js/components/FileInput/FileInput.js

```diff
@@ -0,0 +1,257 @@
+import React, { forwardRef, useContext, useRef } from 'react';
+import styled, { ThemeContext } from 'styled-components';
+
+import { defaultProps } from '../../default-props';
+
+import { disabledStyle, focusStyle, useForwardedRef } from '../../utils';
+
+import { Anchor } from '../Anchor';
+import { Box } from '../Box';
+import { Button } from '../Button';
+import { FormContext } from '../Form/FormContext';
+import { Keyboard } from '../Keyboard';
+import { Text } from '../Text';
+
+import { StyledFileInput } from './StyledFileInput';
+
+// We want the interaction of <input type="file" /> but none of its styling.
+// So, we put what we want to show underneath and
+// position the <input /> on top with an opacity of zero.
+// If there are any files selected, we need to show the buttons to remove them.
+// So, we offset the <input /> from the right by the appropriate width.
+// We don't use Stack because of how we need to control the positioning.
+
+const ContentsBox = styled(Box)`
+  position: relative;
+  ${props => (props.disabled ? disabledStyle() : 'cursor: pointer;')}
+  &:focus {
+    ${focusStyle}
+  }
+  ${props => props.theme.fileInput && props.theme.fileInput.extend};
+  ${props =>
+    props.hover &&
+    props.theme.fileInput &&
+    props.theme.fileInput.hover &&
+    props.theme.fileInput.hover.extend};
+  ${props =>
+    props.dragOver &&
+    props.theme.fileInput &&
+    props.theme.fileInput.dragOver &&
+    props.theme.fileInput.dragOver.extend};
+`;
+
+const Label = styled(Text)`
+  ${props =>
+    props.theme.fileInput &&
+    props.theme.fileInput.label &&
+    props.theme.fileInput.label.extend};
+`;
+
+const Message = styled(Text)`
+  ${props =>
+    props.theme.fileInput &&
+    props.theme.fileInput.message &&
+    props.theme.fileInput.message.extend};
+`;
+
+const FileInput = forwardRef(
+  (
+    {
+      a11yTitle,
+      background,
+      border,
+      disabled,
+      id,
+      renderFile,
+      messages,
+      margin,
+      multiple,
+      name,
+      onChange,
+      pad,
+      value: valueProp,
+      ...rest
```

Comment:
Any way to specify what kind of files can be accepted vs not?

For cases where it's a single file input, any way to give user feedback when they drag multiple files that the input only accepts single file? Found myself wondering why my files weren't being accepted and then realized it was a single file input. -- Maybe this is something to add as a storybook example if it's not something that seems in scope of including in the component.

Reply:
For cases where it's a single file input, any way to give user feedback when they drag multiple files that the input only accepts single file? Found myself wondering why my files weren't being accepted and then realized it was a single file input. -- Maybe this is something to add as a storybook example if it's not something that seems in scope of including in the component.

Reply:
Regarding file types, callers can use the standard `accept` attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept

Reply:
regarding dragging multiple files when `multiple` is not set, we don't have a way to detect that. It's all handled within the `<input type="file" />` element in the DOM. The one event we can tap into is the `onDragOver` event, but that doesn't include anything that indicates how many files are being dragged.

Reply:
Sounds good, thanks for the resources and explanation!